Clean slate

sdvohra opened the pull request:
Added all files to personal repo

sdvohra committed:
Added GraphDB files with simple tracing/perf analysis

sdvohra committed:
Added advanced perf analysis used to generate graphs via chrome://tracing

sdvohra committed:
minor grammar fix in comment

sdvohra committed:
Created test case: creation of transport zones (1) and transport nodes (6)

sdvohra committed:
Bug fix in main method/test case

sdvohra committed:
Basic test case passing for transport zone and transport nodes

sdvohra committed:
Added completed tests for transport zones/ zones and logical ports/switches. Still has bug in clean() method.

sdvohra committed:
Tried various methods to fix clean(), none of which are proving to work. Everything else works as expected.

sdvohra committed:
Added comments/to-do notes for code clean-up. Removed clean method as it is not useful in this context.

sdvohra committed:
Edited classes such that vertices is <UUID, Node> instead of <String, Node>. Still need to revise tests.

sdvohra committed:
Began eliminating Edge class, using UUIDs instead. Not fully implemented.

sdvohra committed:
Completed GraphDB: persistent, uses UUIDs, and removed Edge class fully. Includes passing tests.

sdvohra committed:
Removed AList of Edges in Node class

sdvohra committed:
Added an AppContext class that uses GraphDB - one more layer of abstraction to clean things up; create clean API

sdvohra committed:
Deleted GraphDBHM: implementation of GraphDB with HashMap instead of Corfu object

sdvohra committed:
Deleted Edge file b/c no longer needed

sdvohra committed:
Removed main method from GraphDB, put tests in AppContext instead

sdvohra committed:
Moved all Tracer code to Tracer.java

sdvohra committed:
Code clean-up

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java:
```diff
@@ -43,12 +43,14 @@
 import lombok.extern.slf4j.Slf4j;
 
```
Please remove tracing.

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/GraphDB.java:
```diff
@@ -0,0 +1,149 @@
+
+    public void addEdge(Node from, Node to) {
+        if (from instanceof TransportNode) {
+            ((TransportNode) from).addTransportZoneID(to.getID());
```
What if either 'Node from' or 'Node to' don't exist? You should throw an exception NodeDoesntExist

Also, this method should only operate on a Node type (i.e. remove other types)

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/GraphDB.java:
```diff
@@ -0,0 +1,149 @@
+        to.addEdge(from.getID());
+    }
+
+    public void addEdge(UUID from, UUID to) {
```
Once you address the above comment, then you can remove this method too.

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/GraphDB.java:
```diff
@@ -0,0 +1,149 @@
+
+    /** Returns an iterable of nodes of all vertices adjacent to v. */
+    Iterable<UUID> adjacent(UUID v) {
+        HashSet<UUID> edgeList = vertices.get(v).getEdges();
```
This can throw a null pointer exception if vertices.get returns a null (i.e. v doesnt exist in the map)
Need to check if the node exists in the first place, otherwise throw a NodeDoesntExistException

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/GraphDB.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    /** Clear entire graph: remove all nodes/edges. */
+    void clear() {
```
The graph object should support CRUD (create, read, update and delete operations on Nodes and Edges)

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/Node.java:
```diff
@@ -0,0 +1,74 @@
+ * Created by shriyav on 5/25/17.
+ */
+
+public class Node {
```
Please define a Node and Graph interface. Before you start converting your code and implementing those interfaces, lets meet up and agree on the interface APIs.

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/AppContext.java:
```diff
@@ -0,0 +1,149 @@
+package org.corfudb.samples.graph;
```
split app context into its own class

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/AppContext.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    /** Creates new TransportZone and returns its UUID. */
+    public UUID createTransportZone(String name) {
```
Should be a key type and a key type can be a composite

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/AppContext.java:
```diff
@@ -0,0 +1,149 @@
+        return curr;
+    }
+
+    public void connectTransportNode(UUID tnID, UUID tzID) {
```
Use key instead of string or uuid ?

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/AppContext.java:
```diff
@@ -0,0 +1,149 @@
+         * First, the application needs to instantiate a CorfuRuntime,
+         * which is a Java object that contains all of the Corfu utilities exposed to applications.
+         */
+        CorfuRuntime runtime = getRuntimeAndConnect(corfuConfigurationString);
```
move tests out of the main method. Write junit tests.

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/GraphDB.java:
```diff
@@ -0,0 +1,149 @@
+ */
+
+public class GraphDB {
+    private SMRMap<UUID, Node> vertices;
```
For the sake of uniformity use either a vertex or a node?

Also use a map instead of a smrmap type?

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/GraphDB.java:
```diff
@@ -0,0 +1,149 @@
+    private SMRMap<UUID, Node> vertices;
+    CorfuRuntime rt;
+
+    public GraphDB(CorfuRuntime runtime) {
```
pass in name of graph

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/GraphDB.java:
```diff
@@ -0,0 +1,149 @@
+    public Node getNode(UUID id) { return vertices.get(id); }
+
+    public void addNode(UUID uuid) {
+        vertices.put(uuid, new Node(uuid));
```
what if a node with the same uuid already exists ?
Define exception types? (i.e. NodeAlreadyExistsException)

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/GraphDB.java:
```diff
@@ -0,0 +1,149 @@
+        vertices.put(uuid, new Node(uuid));
+    }
+
+    public void addNode(UUID uuid, String name) {
```
addNode should take Node as a parameter. Maybe define a generic Node type?

Also, remove all other methods that overload that method name?

Maithem commented on samples/src/main/java/org/corfudb/samples/graph/GraphDB.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    public void update(UUID uuid, Map<String, Object> properties) {
+        vertices.get(uuid).getProperties().putAll(properties);
```
update should take a node as a parameter and it should be a put on the vertices map. Since the Node type itself isnt an smr object.

sdvohra committed:
Removed tracing from StreamLogFiles.java

sdvohra committed:
Renamed vertices map to nodes for uniformity

sdvohra committed:
Moved files to correct directories

Commit:
Iterative implementation of DFS/BFS

Commit:
Added basic junit tests in main

Commit:
Clean-up of test in main

Commit:
Add error handling

Commit:
Add CRUD operations

- For both Nodes and edges

Commit:
Add layer of abstraction

Make methods like addNode, addEdge, etc. more general
and applicable to all types of Nodes, not just those
found in L2Workflow (i.e. TransportNode, TransportZone,
etc.).

Commit:
Added JUnit tests

Commit:
Added more JUnit tests

To validate error messages

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/graphdb/GraphDB.java:
```diff
@@ -0,0 +1,178 @@
+
+import java.util.*;
+
+/**
```
Add comprehensive class comment

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/graphdb/GraphDB.java:
```diff
@@ -0,0 +1,178 @@
+    public GraphDB(CorfuRuntime runtime, String n) {
+        nodes = runtime.getObjectsView()
+                .build()
+                .setStreamName("B")     // stream name
```
Name should refer to stream name.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/graphdb/GraphDB.java:
```diff
@@ -0,0 +1,178 @@
+    public void update(UUID uuid, Node n) throws Exception {
+        // Error Handling
+        if (nodes.get(uuid) == null) {
+            throw new Exception("NodeDoesNotExistException");
```
Define your own exception by extending RuntimeException.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/graphdb/GraphDB.java:
```diff
@@ -0,0 +1,178 @@
+        return nodes.size();
+    }
+
+    public void addEdge(Node from, Node to) {
```
check existence

Maithem commented on runtime/src/main/java/org/corfudb/runtime/collections/graphdb/GraphDB.java:
```diff
@@ -0,0 +1,178 @@
+        HashSet<UUID> visited = new HashSet<>();
+        visited.add(first);
+
+        while (!stack.isEmpty()) {
```
try deep graph (the one that was causing the stackoverflow exception

Commit:
Fix GraphDB constructor

String name in GraphDB constructor refers
to stream name, not graph name

Commit:
Renamed addEdge fn in GraphDB to connect

Commit:
Add Node and Graph interface files

No implementation provided yet, however

Commit:
Add comprehensive class comments

Commit:
Error handling in connect fn

Commit:
Improve error handling

- Create 2 java files for exceptions
- Modify tests to account for new exceptions

Commit:
Add changes for directed graph

- Replace edges AList with children and parents
- Add connect and disconnect operations to add
and remove connections between Nodes

Commit:
Fix JUnit tests

= Switch actual and expected in Assert
statements

Commit:
Added deep graph test

Commit:
Added Key interface and descriptions

sdvohra committed:
Restructure GraphDB

- Node class takes an Object class for
simplicity
- Key is no longer necessary
- Node ID determined by Object's hashCode()

sdvohra committed:
Modify/update AppContext and tests

sdvohra committed:
Clean up

sdvohra committed:
More clean-up

- Renamed children/parents of GraphDB
to outward/inward, respectively
- EdgeAlreadyExistsException added to
disallow parallel edges

sdvohra committed:
Revise Graph such that IDs are not leaked

sdvohra committed:
Final clean-up of GraphDB API

- Fixed so that Integer IDs of Nodes are
not leaked
- Test cases failing, however - potential
serializer error

sdvohra committed:
All errors resolved - GraphDB is complete

sdvohra committed:
Clean up of files and comments

sdvohra committed:
Added abstraction: AppContext interface

sdvohra committed:
Moved files to correct location

sdvohra committed:
Added non-native Graph API, both tests pass

sdvohra committed:
Add large test case

- Highlights performance difference between
both implementations

sdvohra committed:
Modified tests/experimenting

sdvohra committed:
Override hashCode to avoid conflicts

sdvohra committed:
Adding basic multithreading

sdvohra committed:
Clean-up of GraphDBAppContextTest

- Make Component class abstract
- Bug fix in GraphDB.addNode()
- Revise test cases

sdvohra committed:
Clean-up of MapsDBAppContext
